sentinel/src/lib/validation.js

```diff
@@ -1,290 +1,356 @@
-     *
-     * OK:
-     *  rule: "regex(\d{5}) && max(11111)"
-     *
-     * If for example you want to validate that patient_id is 5 numbers and it
-     * is unique (or some other custom validation) you need to define two
-     * validation configs/separate rules in your settings. Example validation
-     * settings:
-     *
-     * [
-     *  {
-     *   property: "patient_id",
-     *   rule: "regex(\d{5})",
-     *   message: [{
-     *       content: "Patient ID must be 5 numbers.",
-     *       locale: "en"
-     *   }]
-     *  },
-     *  {
-     *   property: "patient_id",
-     *   rule: "unique('patient_id')",
-     *   message: [{
-     *       content: "Patient ID must be unique.",
-     *       locale: "en"
-     *   }]
-     *  }
-     * ]
-     *
-     * validate() modifies the property value of the second item to
-     * `patient_id_unique` so that pupil.validate() still returns a valid
-     * result.  Then we process the result once more to extract the custom
-     * validation results and error messages.
-     *
-     * @param ignores -
-     *   Optional array keys of doc that is always considered valid
-     *
-     * @callback Array of errors if validation failed, empty array otherwise.
-     * */
-    validate: function(doc, validations, ignores, callback) {
-        var self = module.exports,
-            result = {},
-            errors = [];
+    uniqueWithin: (doc, validation, callback) => {
+      const fields = _.clone(validation.funcArgs);
+      const duration = _parseDuration(fields.pop());
+      const startDate = moment()
+        .subtract(duration)
+        .valueOf();
+      _exists(doc, fields, { startDate: startDate }, (err, result) => {
+        if (err) {
+          logger.error('Error running "uniqueWithin" validation', err);
+        }
+        callback(err, !result);
+      });
+    },
+    exists: (doc, validation, callback) => {
+      const formName = validation.funcArgs[0];
+      const fieldName = validation.funcArgs[1];
+      _exists(
+        doc,
+        [fieldName],
+        { additionalFilter: `form:${formName}` },
+        (err, result) => {
+          if (err) {
+            logger.error('Error running "exists" validation', err);
+          }
+          callback(err, result);
+        }
+      );
+    },
+    /*
+         * Check if the week is a valid ISO week given a year.
+         */
+    isISOWeek: (doc, validation, callback) => {
+      const weekFieldName = validation.funcArgs[0];
+      const yearFieldName =
+        validation.funcArgs.length > 1 ? validation.funcArgs[1] : null;
+      if (
+        !_.has(doc, weekFieldName) ||
+        (yearFieldName && !_.has(doc, yearFieldName))
+      ) {
+        callback(
+          new Error('Validation failed: input field(s) do not exist'),
+          false
+        );
+      } else {
+        const year = yearFieldName
+          ? doc[yearFieldName]
+          : new Date().getFullYear();
```

Comment:
I think Prettier is reformatting this line and then JSHint complains about it. Maybe you could `prettier-ignore` it.

Reply:
👍